src/wp-includes/interactivity-api/class-wp-interactivity-api.php

```diff
@@ -0,0 +1,674 @@
+		if ( ! isset( $this->config_data[ $store_namespace ] ) ) {
+			$this->config_data[ $store_namespace ] = array();
+		}
+		if ( is_array( $config ) ) {
+			$this->config_data[ $store_namespace ] = array_replace_recursive(
+				$this->config_data[ $store_namespace ],
+				$config
+			);
+		}
+		return $this->config_data[ $store_namespace ];
+	}
+
+	/**
+	 * Prints the serialized client-side interactivity data.
+	 *
+	 * Encodes the config and initial state into JSON and prints them inside a
+	 * script tag of type "application/json". Once in the browser, the state will
+	 * be parsed and used to hydrate the client-side interactivity stores and the
+	 * configuration will be available using a `getConfig` utility.
+	 *
+	 * @since 6.5.0
+	 */
+	public function print_client_interactivity_data() {
+		$store      = array();
+		$has_state  = ! empty( $this->state_data );
+		$has_config = ! empty( $this->config_data );
+
+		if ( $has_state || $has_config ) {
+			if ( $has_config ) {
+				$store['config'] = $this->config_data;
+			}
+			if ( $has_state ) {
+				$store['state'] = $this->state_data;
+			}
+			wp_print_inline_script_tag(
+				wp_json_encode(
+					$store,
+					JSON_HEX_TAG | JSON_HEX_AMP
+				),
+				array(
+					'type' => 'application/json',
+					'id'   => 'wp-interactivity-data',
+				)
+			);
+		}
+	}
+
+	/**
+	 * Registers the `@wordpress/interactivity` script modules.
+	 *
+	 * @since 6.5.0
+	 */
+	public function register_script_modules() {
+		wp_register_script_module(
+			'@wordpress/interactivity',
+			includes_url( 'js/dist/interactivity.min.js' )
+		);
+
+		wp_register_script_module(
+			'@wordpress/interactivity-router',
+			includes_url( 'js/dist/interactivity-router.min.js' ),
+			array( '@wordpress/interactivity' )
+		);
+	}
+
+	/**
+	 * Adds the necessary hooks for the Interactivity API.
+	 *
+	 * @since 6.5.0
+	 */
+	public function add_hooks() {
+		add_action( 'wp_footer', array( $this, 'print_client_interactivity_data' ) );
+	}
+
+	/**
+	 * Processes the interactivity directives contained within the HTML content
+	 * and updates the markup accordingly.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param string $html The HTML content to process.
+	 * @return string The processed HTML content. It returns the original content when the HTML contains unbalanced tags.
+	 */
+	public function process_directives( string $html ): string {
+		$p               = new WP_Interactivity_API_Directives_Processor( $html );
+		$tag_stack       = array();
+		$namespace_stack = array();
+		$context_stack   = array();
+		$unbalanced      = false;
+
+		$directive_processor_prefixes          = array_keys( self::$directive_processors );
+		$directive_processor_prefixes_reversed = array_reverse( $directive_processor_prefixes );
+
+		while ( $p->next_tag( array( 'tag_closers' => 'visit' ) ) && false === $unbalanced ) {
+			$tag_name = $p->get_tag();
+
+			if ( 'SVG' === $tag_name || 'MATH' === $tag_name ) {
+				$unbalanced = true;
+				break;
+			}
```

Comment:
I don't think this needs to bail here. Instead, in the `else` condition you can check if `in_array( 'MATH', $tag_stack, true ) || in_array( 'SVG', $tag_stack, true )` or rather [like this](https://github.com/WordPress/performance/blob/4675b7895502eff7d523df94c478134028cf45bd/modules/images/image-loading-optimization/class-ilo-html-tag-processor.php#L289-L296). If so, then you can check if `$p->has_self_closing_flag()`. This will then account for self-closing tags in SVG, like `<path/>` or `<circle/>`.

See how I handled this same thing in my pending Image Loading Optimization plugin:

https://github.com/WordPress/performance/blob/4675b7895502eff7d523df94c478134028cf45bd/modules/images/image-loading-optimization/class-ilo-html-tag-processor.php#L227-L236

When HTML parsing inside of a foreign element (SVG or MathML), then XML processing rules apply.

Reply:
As it seems like SVG and MathML are somewhat tricky ([see](https://wordpress.slack.com/archives/C05NFB818PQ/p1706842893588339)), I'd rather commit this without any support, and then add it carefully on a follow-up.

I've added it to the list of follow-ups in the description.